### ISSUE TITLE
Seemingly endless loop

### ISSUE THREAD

**Issue by Arcnor** [bpodgursky/jbool_expressions#29]
Hi,

First of all, awesome library, thank you very much for making it available!

I've found some issues while trying to parse the (probably too large) following expression with `RuleSet.toDNF(Not.of(<large expression>))`, the library seems to run forever, although I haven't waited more than 1 or 2 minutes:

```
((b4 & b1 & !b3 & !c1 & c2 & c5) | (b4 & b1 & !b3 & !c1 & c2 & !c5) | (b4 & b1 & !b3 & !c1 & !c2 & c5) | (b4 & b1 & !b3 & !c1 & !c2 & !c5) | (b1 & b3 & !c1 & c2 & c5) | (b1 & b3 & !c1 & c2 & !c5) | (b1 & b3 & !c1 & !c2 & c5) | (b1 & b3 & !c1 & !c2 & !c5) | (!b4 & b1 & !b3 & !c1 & c2 & c5) | (!b4 & b1 & !b3 & !c1 & c2 & !c5) | (!b4 & b1 & !b3 & !c1 & !c2 & c5) | (!b4 & b1 & !b3 & !c1 & !c2 & !c5) | (b4 & !b1 & !c1 & b2 & c2 & c5) | (b4 & !b1 & !c1 & b2 & c2 & !c5) | (b4 & !b1 & !c1 & b2 & !c2 & c5) | (b4 & !b1 & !c1 & b2 & !c2 & !c5) | (!b1 & !c1 & !b2 & c2 & c5) | (!b1 & !c1 & !b2 & c2 & !c5) | (!b1 & !c1 & !b2 & !c2 & c5) | (!b1 & !c1 & !b2 & !c2 & !c5) | (!b4 & !b1 & !c1 & b2 & c2 & c5) | (!b4 & !b1 & !c1 & b2 & c2 & !c5) | (!b4 & !b1 & !c1 & b2 & !c2 & c5) | (!b4 & !b1 & !c1 & b2 & !c2 & !c5))
```

**Comment by bpodgursky** [comment]
It took 6hr 40min to run, but I get the answer 

> c1

Obviously this is a ridiculously long time.  I'm not sure how I'd simplify this expression by hand though, so I'm not sure what heuristics this library is missing which could shortcut that answer.  I'm happy to take a look at adding them if you have ideas.

**Comment by Arcnor** [comment]
So it has a solution, nice 😄 

Well, I'm not an expert in this field at all (that's why I was using a library in the first place 😄) but a few things come to mind, hopefully some of them make sense:

1. I'm not sure but it seems this library is not using the Quine–McCluskey algorithm? I this particular case the expression contains 7 variables, which means it will need to find ~4255 prime implicants max. I have no idea about how fast that can be done, I've only skimmed the algorithm.
2. On the Wikipedia page about that algorithm they mention the "Espresso heuristic logic minimizer". Again, no idea about this one, or if it contains better/worse heuristics that you already have.
3. As a naive solution, maybe if you detect an expression is too big, it can be split in N parts and try to minimize those separately, then use the previous results and try to minimize again? You might be doing something like this already, or this might be too naive, but that's the only thing I can think of.

I also need to figure out why I'm generating those massive expressions in the first place, but I'm guessing it cannot be helped, so it will be nice if I can deal with them somehow 😄 

Again, thanks for this nice library!

**Comment by Arcnor** [comment]
Tried with Karnaugh maps as well, and the result was immediate, so maybe that's something to look into when number of variables < X

**Comment by Arcnor** [comment]
So I've adapted some code I found online to calculate Quine-McCluskey and the results are also immediate. This code works for up to 32 variables and in all cases the result was immediate for me.

I'm assuming this project will not be able to work with 32 variables at all, so I think you should use this algorithm, even with this 32 variable limitation. That said, adapting the code to work with >32 variables is more or less trivial, but according to the Wikipedia page on the algorithm, the complexity of the algorithm is exponential to the number of variables used.

I'll comment here when I upload my code somewhere in case you want to adapt it (the Quine McCluskey algorithm is in Java, but the conversion to the AST for this library is in Kotlin, although it's very simple to understand)

**Comment by bpodgursky** [comment]
Got it, thanks.  Yeah, any code would be useful as a reference.  I'll try to take a look at using Quine-McCluskey or the Espresso heuristic minimizer in specific cases sometime in the next couple weeks.  Thanks for digging into this!

**Comment by larswillrich** [comment]
@bpodgursky have you worked on this issue? 
In our project here we're very happy to use this library. Good work! But also we have serious performance problems, because sometimes we're dealing with expressions with a lot of different variables and trying to make a DNF out of it. 
Have you worked on this idea to use the Quine-McCluskey algorithm?

**Comment by bpodgursky** [comment]
I've not been able to get around to this yet.  Might be able to in the next couple weeks but I don't want to make any promises.

To be clear, Quine-McCluskey will help a lot on expressions which are very complex but contain a reasonably small number of variables (I haven't done any performance testing, but 32 sounds like a rough cut point).  So if your expressions have a very large number of variables, they will need different heuristic improvements.

**Comment by Arcnor** [comment]
Chiming in just to say that for the example above (and all others like it), I implemented Quine-McCluskey and it completely fixed the issue.

Right now it's a completely separate step in my code mixed with other things, otherwise I'd share.

**Comment by larswillrich** [comment]
Thanks for the fast reply! 
Where I can find your code @arcnor ?

**Comment by Arcnor** [comment]
It's a private project, I'm afraid, but doing this yourself should not take more than a few hours if I remember correctly, there should be a few implementations online as well that you just need to adapt.

**Comment by larswillrich** [comment]
great work! @bpodgursky 
I'm not sure, but you worked on the simplify method only, right? So you assume to simplify the input argument for the RuleSet.toDNF() method in order to get a faster result?

If yes: Can it may also be possible to integrate this solution somehow in the DNF creation algorithm? Especially to come to a point to maybe reduce the long terms during this process. Do you know what I mean?

**Comment by bpodgursky** [comment]
The method is named simplify, but it's really a toDNF implementation for arbitrary expressions.  I'll clean it up.  And yes, the plan is to integrate it into toDNF.

**Comment by bpodgursky** [comment]
Ok, unless anyone has suggestions or complaints, this is what I'm going to merge and release tonight: https://github.com/bpodgursky/jbool_expressions/pull/38/files#diff-df683d8852f36dcd7c38685dc8b8a58cR24

I'm a little bit torn on how to use this during DNF calculations.  On one hand, QMC is dramatically faster on some bad expressions.  On the other hand, with a reasonable number of variables, it can be a lot slower on really simple expressions -- see my comment for details.  So I'm nervous about merging in a change could slow down simple expression calculations.  So for now I'm just going to add a new method to expose it.  Curious to hear feedback though.

**Comment by bpodgursky** [comment]
Ok, I've released this as 0.20 with the new method RuleSet.toDNFViaQMC as described.  I don't feel great about the naming and I'm open to changing how this is integrated in later releases, but I wanted to get something out for now.  Let me know if you have any concerns or other ideas.
